File: src/stores/identity.ts
import { defineStore } from 'pinia';
import { ref } from 'vue';
import { invoke } from '@tauri-apps/api/core';
import type { DeviceInfo, TrustedDevice } from '../types';

export const useIdentityStore = defineStore('identity', () => {
  const deviceInfo = ref<DeviceInfo | null>(null);
  const qrSvg = ref<string | null>(null);
  const trustedDevices = ref<TrustedDevice[]>([]);
  const error = ref<string | null>(null);
  const isLoading = ref(false);

  async function loadIdentity() {
    isLoading.value = true;
    error.value = null;
    try {
      deviceInfo.value = await invoke<DeviceInfo>('get_device_identity');
    } catch (err) {
      error.value = String(err);
    } finally {
      isLoading.value = false;
    }
  }

  async function loadPairingQr() {
    isLoading.value = true;
    error.value = null;
    try {
      qrSvg.value = await invoke<string>('get_pairing_qr');
    } catch (err) {
      error.value = String(err);
    } finally {
      isLoading.value = false;
    }
  }

  async function loadTrustedDevices() {
    try {
      trustedDevices.value = await invoke<TrustedDevice[]>('list_trusted_devices');
    } catch (err) {
      error.value = String(err);
    }
  }

  async function addTrustedDevice(device: TrustedDevice) {
    try {
      await invoke('add_trusted_device_cmd', { device });
      await loadTrustedDevices();
    } catch (err) {
      error.value = String(err);
      throw err;
    }
  }

  async function removeTrustedDevice(deviceId: string) {
    try {
      await invoke('remove_trusted_device_cmd', { deviceId });
      await loadTrustedDevices();
    } catch (err) {
      error.value = String(err);
      throw err;
    }
  }

  function clearError() {
    error.value = null;
  }

  return {
    deviceInfo,
    qrSvg,
    trustedDevices,
    error,
    isLoading,
    loadIdentity,
    loadPairingQr,
    loadTrustedDevices,
    addTrustedDevice,
    removeTrustedDevice,
    clearError,
  };
});

File: src-tauri/src/commands/identity.rs
use tauri::{AppHandle, Manager, State};

use crate::{
    identity::{
        device_name,
        qr::generate_pairing_qr,
        trusted_devices::{add_trusted_device, remove_trusted_device, save_trusted_devices},
        TrustedDevice,
    },
    AppState,
};

/// Return the public identity of this device (device_id, public key, name).
#[tauri::command]
pub async fn get_device_identity(state: State<'_, AppState>) -> Result<crate::identity::DeviceInfo, String> {
    let identity = state.device_identity.lock().map_err(|e| e.to_string())?;
    match identity.as_ref() {
        Some(id) => Ok(id.device_info(&device_name())),
        None => Err("Device identity not initialised".to_string()),
    }
}

/// Return an SVG QR code encoding the pairing payload for this device.
#[tauri::command]
pub async fn get_pairing_qr(state: State<'_, AppState>) -> Result<String, String> {
    let identity = state.device_identity.lock().map_err(|e| e.to_string())?;
    match identity.as_ref() {
        Some(id) => generate_pairing_qr(&id.device_info(&device_name())),
        None => Err("Device identity not initialised".to_string()),
    }
}

/// Return the list of trusted (paired) devices.
#[tauri::command]
pub async fn list_trusted_devices(state: State<'_, AppState>) -> Result<Vec<TrustedDevice>, String> {
    Ok(state
        .trusted_devices
        .lock()
        .map(|d| d.clone())
        .unwrap_or_default())
}

/// Add a device to the trusted list and persist the change.
#[tauri::command]
pub async fn add_trusted_device_cmd(
    device: TrustedDevice,
    state: State<'_, AppState>,
    app: AppHandle,
) -> Result<(), String> {
    let mut devices = state.trusted_devices.lock().map_err(|e| e.to_string())?;
    add_trusted_device(&mut devices, device);
    let data_dir = app.path().app_data_dir().map_err(|e| e.to_string())?;
    save_trusted_devices(&data_dir, &devices)
}

/// Remove a device from the trusted list by its `device_id` and persist.
#[tauri::command(rename_all = "camelCase")]
pub async fn remove_trusted_device_cmd(
    device_id: String,
    state: State<'_, AppState>,
    app: AppHandle,
) -> Result<(), String> {
    let mut devices = state.trusted_devices.lock().map_err(|e| e.to_string())?;
    remove_trusted_device(&mut devices, &device_id);
    let data_dir = app.path().app_data_dir().map_err(|e| e.to_string())?;
    save_trusted_devices(&data_dir, &devices)
}
